### PR TITLE
nixos/xdg: disable portals (again, again)

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
See the commits and hopefully you'll get the gist of the story :rofl: 

I'm not sure if the assertion in flatpak is the best way to go, but I'd like
these things to play nice together, and most certainly not default enable `GTK_USE_PORTAL`
because that will affect non-flatpak applications.

Additionally, having an option like `gtkUsePortal` which just sets
```
environment.variables.GTK_USE_PORTAL = "1";
```

is slightly overkill in my opinion. A usecase that would better belong in the NixOS wiki.

However, many other distros are getting support requests to have this for the native filechoosers, so this would be an easy way to have an opt-in within NixOS. 
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/config/xdg/portal.nix [hunk 1]
```diff
@@ -28,10 +40,17 @@ with lib;
 
     in mkIf cfg.enable {
 
+      assertions = [
+        { assertion = (cfg.gtkUsePortal -> cfg.extraPortals != []);
+          message = "Setting xdg.portal.gtkUsePortal to true without a XDG portal implementation in xdg.portal.extraPortals will cause things to break.";
```
Comment (samueldr) [inline nixos/modules/config/xdg/portal.nix]: Could we nudge the user towards an option or a solution, instead of only admonishing them for their choice?

**Review by samueldr — COMMENTED** [review]
Minor change for the assert, otherwise looks good.

**Review by samueldr — APPROVED** [review]

**Comment by worldofpeace** [comment]
cc @michaelpj (might be interested in reviewing)

**Comment by michaelpj** [comment]
Could we have `GTK_USE_PORTAL` just be conditional on `xdg.extraPortals` being non-empty? That way we avoid another option.

Also, what's the rationale for having this off by default? All the other XDG things are on by default. Arguably they shouldn't be, but I'd prefer consistency if possible.

**Comment by samueldr** [comment]
One of the reason for being off by default is that it increases the base closure size quite a bit. I think it was 300MiB?

It, at least, made the sd_image too big to generate, blocking channels.

The reason is that it brings some gnomey bits into the non-graphical images from having the option turned on by default.

**Comment by michaelpj** [comment]
Hmm, that seems strange but I have no idea what's in this portal package. I don't suppose we can just configure it to pull in less by default?

**Comment by worldofpeace** [comment]
`xdg-desktop-portal` was only factored outside of `flatpak` because of the usecase in https://github.com/NixOS/nixpkgs/pull/65449/commits/4cb9a1fa7d275b230ccb6859aecf5794a06d93c8.

Since it's directly related to flatpak it's pulling in all those dependencies, it doesn't belong default.
The name `xdg` doesn't really have relation to the other xdg modules which are just for freedesktop specs.
And I think the closure was 604MiB.

> Could we have GTK_USE_PORTAL just be conditional on xdg.extraPortals being non-empty? That way we avoid another option.

See description https://github.com/NixOS/nixpkgs/pull/65449/commits/4cb9a1fa7d275b230ccb6859aecf5794a06d93c8. Really shouldn't be set default like that.

**Comment by michaelpj** [comment]
Okay, makes sense to me.

Should the desktop environments set the GKT option?

**Comment by michaelpj** [comment]
Oh also, could we maybe put a little more of this reasoning into comments in the file, it can be a little hard to discover these things later otherwise!

**Comment by worldofpeace** [comment]
> Okay, makes sense to me.
> 
> Should the desktop environments set the GKT option?

At the moment only GNOME should do that since it was designed for that environment.
And I think that's something that should be left up to the user, unless it's documented that upstream in that DE wants that.


> Oh also, could we maybe put a little more of this reasoning into comments in the file, it can be a little hard to discover these things later otherwise!

Which in particular, I feel like the options descriptions and assertions suffice?

**Comment by michaelpj** [comment]
Personally, I'd add a comment above the `enable` option explaining why it's disabled by default. I'd maybe add something to the option description for the GTK option saying what you just said (i.e. DE modules should enable this if it's advised by upstream).

**Comment by worldofpeace** [comment]
> Personally, I'd add a comment above the enable option explaining why it's disabled by default.

Perhaps I could do that because it implicates flatpak and a graphical environment heavily.

> add something to the option description for the GTK option saying what you just said (i.e. DE modules should enable this if it's advised by upstream).

Huh, somehow in the previous comment I thought you meant add `xdg-desktop-portal-gtk` to `extraPortals` :smile:.

The actual description of `gtkUsePortal` has
```
Defaults to <literal>false</literal> to respect its opt-in nature.
```

Only thing I can think of would be the pointer for Firefox as an example.

**Comment by michaelpj** [comment]
Well, at the moment it's still pretty unclear to me when to enable `gtkUsePortal`. Perhaps that's fine and it's one of those "you'll know when you need it" options. But it would be nice to have a little bit more.

**Comment by abbradar** [comment]
I use GNOME 3 on some machines and they too had issues with Firefox open file dialog. Also GNOME was very slow to start a session, as if something timeouted. Given that maybe we should disable it for GNOME by default?

**Comment by worldofpeace** [comment]
> I use GNOME 3 on some machines and they too had issues with Firefox open file dialog.

That is no longer a issue

> Also GNOME was very slow to start a session, as if something timeouted. Given that maybe we should disable it for GNOME by default?

I think I've noticed this in https://github.com/NixOS/nixpkgs/pull/65291 so perhaps there was an issue before this. At any rate portals should be enabled for GNOME, so that is just an issue we should be able to fix in #65291.

**Pull request merged**